js/modules/k6/data/data.go

```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package data
+
+import (
+	"context"
+
+	"github.com/dop251/goja"
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/js/internal/modules"
+	"github.com/loadimpact/k6/lib"
+	"github.com/pkg/errors"
+)
+
+type data struct{}
+
+func init() {
+	modules.Register("k6/data", new(data))
+}
+
+const sharedArrayNamePrefix = "k6/data/SharedArray."
+
+// XSharedArray is a constructor returning a shareable read-only array
+// indentified by the name and having their contents be whatever the call returns
+func (d *data) XSharedArray(ctx context.Context, name string, call goja.Callable) (goja.Value, error) {
+	if lib.GetState(ctx) != nil {
+		return nil, errors.New("new SharedArray must be called in the init context")
+	}
+
+	initEnv := common.GetInitEnv(ctx)
+	if initEnv == nil {
+		return nil, errors.New("missing init environment")
+	}
+	if len(name) == 0 {
+		return nil, errors.New("empty name provided to SharedArray's constructor")
+	}
+
+	name = sharedArrayNamePrefix + name
+	value := initEnv.SharedObjects.GetOrCreateShare(name, func() interface{} {
+		return getShareArrayFromCall(common.GetRuntime(ctx), call)
+	})
+	array, ok := value.(sharedArray)
+	if !ok { // TODO more info in the error?
+		return nil, errors.New("wrong type of shared object")
+	}
+
+	return array.wrap(&ctx, common.GetRuntime(ctx)), nil
+}
+
+func getShareArrayFromCall(rt *goja.Runtime, call goja.Callable) sharedArray {
+	gojaValue, err := call(goja.Undefined())
+	if err != nil {
+		common.Throw(rt, err)
+	}
+	obj := gojaValue.ToObject(rt)
+	if obj.ClassName() != "Array" {
+		common.Throw(rt, errors.New("only arrays can be made into SharedArray")) // TODO better error
+	}
+	arr := make([]string, obj.Get("length").ToInteger())
+
+	cal, err := rt.RunString(`(function(input, output) {
```

Comment:
Can you explain (and maybe add as a code comment here) why we're encoding the data to JSON on the JS side? `JSON.stringify()` does something magical that `json.Marshal()` doesn't? Or does this save us from utf8<->utf16 conversions?

Reply:
The commit comment explains it :P. Using JSON.parse was the easiest way to get a goja object thing that can then be frozen as otherwise `json.Decode` gets you a golang map which ... just isn't a goja object. You can make it into a value but it still the golang map underneath and doesn't let itself be frozen (the goja author actually commented on that). 
So at that point, I had the question - do I use two different libraries to do the encode/decode or not, and went with "not" ;).

They have some ... interesting [behavior differences around stuff like functions and other strange types](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify).

But IMO it is much better to be as JS like as possible - explaining that what we do is JSON.parse/JSON.stringify a value instead of json.Encode/Decode in golang and that is why your strange value doesn't work will be again easier for users. Also possibly make it possible for them to make it work.

Another benefit is this way I iterate only once here instead of twice, or once but use string indexes and `Object.Get` :scream:.

Reply:
Please add a code comment with that, since I doubt anyone will read the commit description in the future and might try to "optimize" it

Reply:
That's the problem .. people don't read git commits :sob: